homeassistant/components/repetier/sensor.py

```diff
@@ -37,11 +47,15 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
         temp_id = info["temp_id"]
         name = info["name"]
         if temp_id is not None:
-            name = "{}{}{}".format(name, SENSOR_TYPES[sensor_type][3], temp_id)
+            _LOGGER.debug("%s Temp_id: %s", sensor_type, ent_id[sensor_type])
+            name = "{}{}{}".format(
+                name, SENSOR_TYPES[sensor_type][3], ent_id[sensor_type]
+            )
+            ent_id[sensor_type] = ent_id[sensor_type] + 1
         else:
             name = "{}{}".format(name, SENSOR_TYPES[sensor_type][3])
         sensor_class = sensor_map[sensor_type]
-        entity = sensor_class(api, temp_id, name, printer_id, sensor_type)
+        entity = sensor_class(api, ent_id[sensor_type], name, printer_id, sensor_type)
```

Comment:
Just out of curiosity, isn't it possible to store the correct `temp_id` up there in `_load_entities`? It would make this field much easier to understand. And we wouldn't need the `ent_id` map.

Reply:
That's how it have been until now - that didn't work exactly as supposed, hence this rewrite.

Reply:
So `temp_id` is wrong, isn't it? Why not store the correct id?

Reply:
The correct value is set to temp_id on every sensor, but for some reason only the last one is stored.
I have been looking at this for a few days, but can't see why it shouldn't work the way it was - but it isn't.
If you have an idea to make it work, using temp_id rather than the ugly mapping, please do show.

Reply:
Just did a quick debug, which doesn't quite make sense....

```
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] Sensor bed_temperature is idx 0
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] Sensor extruder_temperature is idx 0
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] Sensor extruder_temperature is idx 1
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] Sensor chamber_temperature is idx 0
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] bed_temperature created with idx 0 in array 0
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] extruder_temperature created with idx 1 in array 1
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] extruder_temperature created with idx 1 in array 2
2019-11-14 19:14:50 DEBUG (SyncWorker_8) [homeassistant.components.repetier] chamber_temperature created with idx 0 in array 3

```

It sees the correct index for the sensor, but when created both extruders are created with index 1

Reply:
It's somewhere in here:

```
    def _load_entities(self):
        sensor_info = []
        for pidx, printer in enumerate(self.printers):
            for sensor_type in self.sensors:
                info = {}
                info["sensor_type"] = sensor_type
                info["printer_id"] = pidx
                info["name"] = printer.slug
                info["printer_name"] = self.conf_name


                known = f"{printer.slug}-{sensor_type}"
                if known in self._known_entities:
                    continue


                methods = API_PRINTER_METHODS[sensor_type]
                if "temp_data" in methods["state"].values():
                    prop_data = getattr(printer, methods["attribute"])
                    if prop_data is None:
                        continue
                    for idx, _ in enumerate(prop_data):
                        _LOGGER.debug("%s idx %s", sensor_type, idx)
                        info["temp_id"] = idx
                        sensor_info.append(info)
                else:
                    info["temp_id"] = None
                    sensor_info.append(info)


                self._known_entities.add(known)


        if not sensor_info:
            return
        load_platform(self._hass, "sensor", DOMAIN, sensor_info, self.config)
```

Reply:
Some thoughts:
1. You're iterating all over the printers here:
    `for pidx, printer in enumerate(self.printers):`
2. You're iterating over all sorts of sensors of one printer here:
    `for sensor_type in self.sensors:`
3. You're iterating over all sensors of one kind here:
    `for idx, _ in enumerate(prop_data):`

And you're saving the last found sensor into the `info` dict. So only the last sensor of one kind is actually saved, while all other sensors are overwritten by the last one. Am I right?

Reply:
3. iterating over temperature sensors to look for multiples (ie. extruders)
Only the temperature sensors fails - info[temp_id] in the temperature sensors dict is overwritten with the last index - but both are actually created as they are supposed to, only the info[temp_id] is overwritten.

Reply:
Did you debug through your code? I mean using the debug mode and breakpoints? It will become clear then, that both extruders are created with temp_id = 1

Reply:
As seen in the log i posted earlier, the sensors info[temp_id] is set correctly, but after they are created both extruders are suddenly index 1.

Still learning Python, and I write the code in Visual Studio 2017 (which I'm fairly familiar with, coding in C#), can't see how I can use breakpoints from there as the code is run from the test server and not from within Visual Studio

Reply:
Actually, if I change

```
                    for idx, _ in enumerate(prop_data):
                        _LOGGER.debug("Sensor %s is idx %s", sensor_type, idx)
                        info["temp_id"] = idx
                        sensor_info.append(info)
```

to

```
                    for idx, _ in enumerate(prop_data):
                        _LOGGER.debug("Sensor %s is idx %s", sensor_type, idx)
                        info["temp_id"] = idx
                        sensor_info.append(info)
                        info["temp_id"] = None
```

both sensors temp_id are set to None - which I absolutely don't understand.

Reply:
https://developers.home-assistant.io/docs/en/development_environment.html#developing-with-devcontainer

Reply:
To make this clear, info is still the same object since you initialized it. When you add it to the list it will be referenced there, but nothing more. It won't be copied or sth like that.

I did some tests:
```python
>>> sensors = []
>>> sensor = {}
>>> sensor["one"] = 1
>>> sensor["two"] = 2
>>> sensors.append(sensor)
>>> print(sensor)
{'one': 1, 'two': 2}
>>> print(sensors)
[{'one': 1, 'two': 2}]
>>> sensor["one"] = 3
>>> print(sensor)
{'one': 3, 'two': 2}
>>> print(sensors)
[{'one': 3, 'two': 2}]
>>> 
```

Reply:
```
2019-11-15 18:25:27 DEBUG (SyncWorker_5) [homeassistant.components.repetier] Sensor extruder_temperature is idx 0
> /home/repetier/home-assistant/homeassistant/components/repetier/__init__.py(254)_load_entities()
-> breakpoint()
(Pdb) print(info["temp_id"])
0
(Pdb) c
2019-11-15 18:25:42 DEBUG (SyncWorker_5) [homeassistant.components.repetier] Sensor extruder_temperature is idx 1
> /home/repetier/home-assistant/homeassistant/components/repetier/__init__.py(255)_load_entities()
-> sensor_info.append(info)
(Pdb) print(info["temp_id"])
1
(Pdb) c

```

Reply:
> As you can see, even though I've changed sensor **after** appending it to sensors, it changed the object within the list as well since it's still the same object. That wasn't clear to me as well.

Hmm - that's odd, I was under the impression it was copied into sensors, just like in .NET and PHP.
That explains this issue pretty clearly, then I don't how we could work around the ent_id mapping thing.
Perhaps leave this be as it is for now, and then go back figuring out how to make this prettier in a future PR?!

Reply:
Just copy the dict when you need a new one:
https://stackoverflow.com/questions/2465921/how-to-copy-a-dictionary-and-only-edit-the-copy

Reply:
Done